File: 2023/src/main/java/uk/co/jonathonhenderson/aoc/common/Direction.java
package uk.co.jonathonhenderson.aoc.common;

public enum Direction {
  NORTH,
  EAST,
  SOUTH,
  WEST;

  public Point delta() {
    return switch (this) {
      case NORTH -> Point.of(0, -1);
      case EAST -> Point.of(1, 0);
      case SOUTH -> Point.of(0, 1);
      case WEST -> Point.of(-1, 0);
    };
  }

  public Direction opposite() {
    return switch (this) {
      case NORTH -> SOUTH;
      case EAST -> WEST;
      case SOUTH -> NORTH;
      case WEST -> EAST;
    };
  }
}

File: 2019/src/main/kotlin/com/github/jrhenderson1988/adventofcode2019/common/Direction.kt
package com.github.jrhenderson1988.adventofcode2019.common

enum class Direction(val delta: Pair<Int, Int>) {
    NORTH(Pair(0, -1)),
    SOUTH(Pair(0, 1)),
    WEST(Pair(-1, 0)),
    EAST(Pair(1, 0));

    fun nextPoint(current: Pair<Int, Int>) = Pair(current.first + delta.first, current.second + delta.second)

    fun opposite() = when (this) {
        NORTH -> SOUTH
        SOUTH -> NORTH
        EAST -> WEST
        WEST -> EAST
    }

    companion object {
        fun neighboursOf(current: Pair<Int, Int>) = values().map { it.nextPoint(current) }
    }
}